src/cli/argsource.rs

```diff
@@ -0,0 +1,54 @@
+/// Abstracts over reading the current process environment variables as a
+/// zero-cost abstraction to support threaded in-process testing.
+use std::env;
+
+pub trait ArgSource {
+    type Args: Iterator<Item = String>;
+
+    fn args(&self) -> Self::Args;
+}
+
+/// Implements ArgSource with `std::env::args`
+pub struct OsArgs {}
+
+impl ArgSource for OsArgs {
+    type Args = env::Args;
+    fn args(&self) -> Self::Args {
+        env::args()
+    }
+}
+
+/// Helper for ArgSource over `Vec<String>`
+pub struct VecArgs {
+    v: Vec<String>,
+    i: usize,
+}
+
+impl From<&Vec<String>> for VecArgs {
```

Comment:
Would it be okay to have this as `From<&[String]>` and to have a complementary `From<&[&str]>` perhaps?  I tried to persuade an `impl<S: AsRef<str>> From<&[S]> for Foo` to exist, but it started whining that `&[&str; 2]` wasn't compatible which made me a bit sad.

Reply:
.collect() on those should do the right thing for us already; this type is *much* easier as an owned type. I mean yes - but in my conversions in run() in the -wip branch, it isn't an issue.